### PR TITLE
Improved AI chooseAttackTarget

### PR DESCRIPTION
**Pull request by tuvus**
This commit restructures the AI's chooseAttackTarget to assign a value to each tile that can be attacked and choose the best one. This style should be easier for future developers to tweak and edit. It may also be helpful later when we want to say chooseAttackTarget with value > 100 or else heal or some more complex military action instead of having a series of if statements.

Military units are now more inclined to attack Cities by default. The more allied units around the city, the higher the priority. Siege units now almost fully prioritize cities (as that is what they are meant to do).

### CODE REVIEW

File: core/src/com/unciv/logic/automation/unit/BattleHelper.kt [hunk 1]
```diff
@@ -52,64 +54,101 @@ object BattleHelper {
...
+            // Associate enemy units with number of hits from this unit to kill them
+            val attacksToKill = (militaryUnit.health.toFloat() / 
+                BattleDamage.calculateDamageToDefender(MapUnitCombatant(attacker), MapUnitCombatant(militaryUnit))).coerceAtLeast(1f)
+            // We can kill them in this turn
+            if (attacksToKill <= 1) attackValue += 100
+            // On average, this should take around 3 turns, so -30
+            else attackValue -= (attacksToKill * 10).toInt()
+        } else {
+            val civilianUnit = attackTile.tileToAttack.civilianUnit
+            if (civilianUnit != null) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/unit/BattleHelper.kt:135]: Please babe this 'else if', to reduce nesting - you can set the val inside the else if

File: core/src/com/unciv/logic/automation/unit/BattleHelper.kt [hunk 2]
```diff
@@ -52,64 +54,101 @@ object BattleHelper {
...
+                attackValue = 0
+                attackValue += 50
+                // Only melee units should really attack/capture civilian units, ranged units take more than one turn
+                if (attacker.baseUnit.isMelee()) {
+                    if (civilianUnit.isGreatPerson()) {
+                        attackValue += 50
+                        // This is really good if we can kill a great general
+                        if (civilianUnit.isGreatPersonOfType("Great General")) attackValue += 150
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/unit/BattleHelper.kt:143]: In general, all great people should be prioritized

File: core/src/com/unciv/logic/automation/unit/BattleHelper.kt [hunk 3]
```diff
@@ -52,64 +54,101 @@ object BattleHelper {
...
+                    }
+                    if (civilianUnit.hasUnique(UniqueType.FoundCity)) attackValue += 30
+                }
+            }
         }
+        // Prioritise closer units as they are generally more threatening to this unit
+        attackValue += (attackTile.movementLeftAfterMovingToAttackTile * 5).toInt()
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/unit/BattleHelper.kt:150]: Only true for military units, no? That way both sides of the 'if' can return immediately, as well - better for readability

File: core/src/com/unciv/logic/automation/unit/BattleHelper.kt [hunk 4]
```diff
@@ -52,64 +54,101 @@ object BattleHelper {
...
+    /**
+     * Returns a value which represents the attacker's motivation to attack a unit.
+     */
+    private fun getUnitAttackValue(attacker: MapUnit, attackTile: AttackableTile): Int {
+        // Base attack value, there is nothing there...
+        var attackValue = Int.MIN_VALUE
+        // Prioritize attacking military
+        val militaryUnit = attackTile.tileToAttack.militaryUnit
+        if (militaryUnit != null) {
+            attackValue = 0
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/unit/BattleHelper.kt:124]: =150 directly

File: core/src/com/unciv/logic/automation/unit/BattleHelper.kt [hunk 5]
```diff
@@ -52,64 +54,101 @@ object BattleHelper {
...
 
+    /**
+     * Choses the best target in attackableEnemies, this could be a city or a unit.
+     */
     private fun chooseAttackTarget(unit: MapUnit, attackableEnemies: List<AttackableTile>): AttackableTile? {
-        val cityTilesToAttack = attackableEnemies.filter { it.tileToAttack.isCityCenter() }
-        val nonCityTilesToAttack = attackableEnemies.filter { !it.tileToAttack.isCityCenter() }
-
+        var highestAttackValue = 0
+        var attackTile: AttackableTile? = null
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/unit/BattleHelper.kt:62]: This whole thing feels like maxBy {}

File: core/src/com/unciv/logic/automation/unit/BattleHelper.kt (hunk 2 above)
Comment (tuvus) [reply core/src/com/unciv/logic/automation/unit/BattleHelper.kt:143]: Do you mean 'UniqueType.GreatGeneralProvidesDoubleCombatBonus'? It doesn't sound right though, other than that, I can't find any other Unique to match it.

File: core/src/com/unciv/logic/automation/unit/BattleHelper.kt (hunk 3 above)
Comment (tuvus) [reply core/src/com/unciv/logic/automation/unit/BattleHelper.kt:150]: I don't think so. The idea is, that if the Unit has to move less to attack, it is closer to its friends and therefore, less vulnerable.

Comment (trentonk23) [reply core/src/com/unciv/logic/automation/unit/BattleHelper.kt:145]: NEEDS THIS

Comment (yairm210) [reply core/src/com/unciv/logic/automation/unit/BattleHelper.kt:145]: Ah good point

**Review by yairm210 — APPROVED** [review]

**Pull request merged**